### PR TITLE
Add C#-style type accelerators and suffixes for ushort, uint, ulong, and short literals.

### PR DESCRIPTION
**Pull request by vexx32**
## PR Summary

This has been split off of #7509 at request of @iSazonov as that PR has become overly complex, and this portion of the PR was already approved by Powershell Committee.

Adds the following numeric literal suffixes:

* 'u' suffix (uint32)
* 's' suffix (short)
* 'ul' suffix (ulong)
* 'us' suffix (ushort)

Adds the following type accelerators:

* `[short]`
* `[ushort]`
* `[uint]`
* `[ulong]`

And of course, adds tests for all the above.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

**Comment by vexx32** [comment]
I don't intend to fix the code factor issues here. I'll address them during the parser refactor that will come after this.

File: src/System.Management.Automation/engine/parser/tokenizer.cs [hunk 1]
```diff
@@ -3268,13 +3278,33 @@ private static bool TryGetNumberValue(string strNum, bool hex, bool real, char s
...
+                                }
+                                else
+                                {
+                                    result = ((uint)Convert.ChangeType(d, typeof(uint), CultureInfo.InvariantCulture) * (uint)multiplier);
+                                }
                             }
                             else
                             {
-                                result = d * multiplier;
+                                result = d;
```
[the remark this answers is not in the document]
Comment (vexx32) [reply src/System.Management.Automation/engine/parser/tokenizer.cs]: Ah, I thought I got all those. Correcting...

(I was trying to see if I could do the multiplier before the conversions to avoid throwing too much here, but it doesn't work *because* Int64.MaxValue is too high to retain precision with Double values in the upper range there.)

File: src/System.Management.Automation/engine/parser/tokenizer.cs [hunk 2]
```diff
@@ -3268,13 +3278,33 @@ private static bool TryGetNumberValue(string strNum, bool hex, bool real, char s
...
                                 d = -0.0;
                             }
-                            if (suffix == 'l' || suffix == 'L')
+
+                            if (suffix == NumberSuffixFlags.Long)
                             {
-                                result = ((long)Convert.ChangeType(d, typeof(long), CultureInfo.InvariantCulture)) * multiplier;
+                                result = ((long)Convert.ChangeType(d, typeof(long), CultureInfo.InvariantCulture) * multiplier);
+                            }
+                            else if (suffix == NumberSuffixFlags.Short)
```
[the remark this answers is not in the document]
Comment (vexx32) [reply src/System.Management.Automation/engine/parser/tokenizer.cs]: Hmmm. Agreed!

File: src/System.Management.Automation/engine/parser/tokenizer.cs [hunk 3]
```diff
@@ -3304,6 +3334,33 @@ private static bool TryGetNumberValue(string strNum, bool hex, bool real, char s
                         result = null;
                         return false;
                     }
+                    else if (suffix == NumberSuffixFlags.Short)
+                    {
+                        if (short.TryParse(strNum, style, NumberFormatInfo.InvariantInfo, out short s))
+                        {
+                            result = s * (short)multiplier;
```
[the remark this answers is not in the document]
Comment (vexx32) [reply src/System.Management.Automation/engine/parser/tokenizer.cs]: At present, that will throw an error and is expected. 

With some of the upcoming changes (that you can see if you build the version in the original PR) that instead simply registers as an invalid numeric literal, as I am able to verify first that the multiplier does not overflow the intended value type bounds without introducing a morass of if/else.

File: src/System.Management.Automation/engine/parser/tokenizer.cs [hunk 4]
```diff
@@ -3304,6 +3334,33 @@ private static bool TryGetNumberValue(string strNum, bool hex, bool real, char s
...
+                            return true;
+                        }
+                        result = null;
+                        return false;
+                    }
+                    else if (suffix.HasFlag(NumberSuffixFlags.Unsigned) && ulong.TryParse(strNum, style, NumberFormatInfo.InvariantInfo, out ulong u))
+                    {
+                        u *= (ulong)multiplier;
```
[the remark this answers is not in the document]
Comment (vexx32) [reply src/System.Management.Automation/engine/parser/tokenizer.cs]: I am also pondering that myself, but I've not seen a completely necessary need to change it. It does seem to simplify calculations against other signed types, however; unsigned types often need a cast when doing multiplication with signed types.

File: test/powershell/Language/Parser/Parser.Tests.ps1 [hunk 5]
```diff
@@ -660,7 +660,13 @@ foo``u{2195}abc
 			@{ Script = "0x80000000"; Expected = $([int32]::MinValue) }
 			@{ Script = "0xFFFFFFFF"; Expected = "-1" }
 			@{ Script = "0x7fffffff"; Expected = $([int32]::MaxValue) }
-			@{ Script = "0x100000000"; Expected = [int64]0x100000000 }
+            @{ Script = "0x100000000"; Expected = [int64]0x100000000 }
+            #Tests for short notation
+            @{ Script = "10s"; Expected = "10" }
```
[the remark this answers is not in the document]
Comment (vexx32) [reply test/powershell/Language/Parser/Parser.Tests.ps1]: That is probably wise. I'll take some time over lunch to add tests for value types. :)

File: test/powershell/Language/Parser/Parser.Tests.ps1 [hunk 6]
```diff
@@ -686,6 +692,23 @@ foo``u{2195}abc
             param ( $Script, $Expected )
             ExecuteCommand $Script | Should -Be $Expected
         }
+
+        $testInvalidNumerals = @(
+            @{ Script = "16p" }
+            @{ Script = "20ux" }
```
[the remark this answers is not in the document]
Comment (vexx32) [reply test/powershell/Language/Parser/Parser.Tests.ps1]: Sure, if we want to. `x` is neither a valid type suffix nor a multiplier in PS, though, so testing for it is more or less the same as testing `16p` at the moment. We could go through the alphabet if you want. :)

File: test/powershell/Language/Parser/Parser.Tests.ps1 [hunk 7]
```diff
@@ -686,6 +692,23 @@ foo``u{2195}abc
...
+            @{ Script = "18uu" }
+            @{ Script = "21ss" }
+            @{ Script = "100ll" }
+            @{ Script = "150su" }
+            @{ Script = "10ds" }
```
[the remark this answers is not in the document]
Comment (vexx32) [reply test/powershell/Language/Parser/Parser.Tests.ps1]: Ah, good catch. Absolutely.

File: src/System.Management.Automation/engine/parser/tokenizer.cs [hunk 8]
```diff
@@ -3304,6 +3334,33 @@ private static bool TryGetNumberValue(string strNum, bool hex, bool real, char s
                         result = null;
                         return false;
                     }
+                    else if (suffix == NumberSuffixFlags.Short)
```
[the remark this answers is not in the document]
Comment (vexx32) [reply src/System.Management.Automation/engine/parser/tokenizer.cs]: Again, agreed. Sorted out. :)

File: src/System.Management.Automation/engine/parser/tokenizer.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (iSazonov) [reply src/System.Management.Automation/engine/parser/tokenizer.cs]: Closed under next PR.

File: src/System.Management.Automation/engine/parser/tokenizer.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (iSazonov) [reply src/System.Management.Automation/engine/parser/tokenizer.cs]: Closed.

File: test/powershell/Language/Parser/Parser.Tests.ps1 (hunk 7 above)
[the remark this answers is not in the document]
Comment (vexx32) [reply test/powershell/Language/Parser/Parser.Tests.ps1]: @iSazonov Also noting down now before I forget:
Presently the literal `0x10d` is *legal* in PowerShell, although `Decimal.TryParse()` does *not* allow the hexadecimal syntax and that parse fails. Its actual return type is `Int32`. 

I plan on fixing this as I go through the parser rewrite after this PR.

[the remark this answers is not in the document]
Comment (iSazonov) [reply test/powershell/Language/Parser/Parser.Tests.ps1]: 0x10d - have we tests for it?

File: test/powershell/Language/Parser/Parser.Tests.ps1 (hunk 6 above)
[the remark this answers is not in the document]
Comment (iSazonov) [reply test/powershell/Language/Parser/Parser.Tests.ps1]: No :-) We should cover _all_ code paths. Do we cover?

File: test/powershell/Language/Parser/Parser.Tests.ps1 (hunk 7 above)
[the remark this answers is not in the document]
Comment (vexx32) [reply test/powershell/Language/Parser/Parser.Tests.ps1]: Oh, lol. I am clearly not *quite* up to speed this morning. Of *course* 0x10d is legal. D is a hexadecimal character... and that's exactly why decimals will never result from a hex literal ending in `d`. I'm gonna make a comment in the code that reflects this, because it's not immediately obvious why Decimal doesn't even bother to check for hex literals.

[the remark this answers is not in the document]
Comment (vexx32) [reply test/powershell/Language/Parser/Parser.Tests.ps1]: (and yes I believe there are tests for it, but I'll pop one in the type checks just to be sure.)

[the remark this answers is not in the document]
Comment (vexx32) [reply test/powershell/Language/Parser/Parser.Tests.ps1]: Added all these. :)

**Comment by vexx32** [comment]
OK, pretty sure I got that short parsing sorted. I can understand doing operations in the console types can be freely converted, but if you're gonna specify `s` and get an int32, there's not much of a point!

**Comment by rjmholt** [comment]
@iSazonov I've added the review tag back, since @BrucePay added back in the other PR.

I think the scope of these changes are larger than the committee originally thought, and since they are pretty fundamental changes to some core functionality, they probably deserve some scrutiny and a holistic consideration.

File: src/System.Management.Automation/engine/parser/tokenizer.cs [hunk 9]
```diff
@@ -3255,6 +3255,8 @@ private static bool TryGetNumberValue(string strNum, bool hex, bool real, Number
                 {
                     NumberStyles style = NumberStyles.AllowLeadingSign | NumberStyles.AllowDecimalPoint |
                                          NumberStyles.AllowExponent;
+
+                    // Decimal parser does not accept hex literals, and 'd' is a valid hex character, so will never be read as Decimal literal
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/parser/tokenizer.cs]: We could add sample.

File: test/powershell/Language/Parser/Parser.Tests.ps1 [hunk 10]
```diff
@@ -709,6 +711,26 @@ foo``u{2195}abc
...
+            @{ Script = "0x10ul"; Expected = "UInt64" }
+            @{ Script = "0x10us"; Expected = "UInt16" }
+            @{ Script = "0x10s"; Expected = "Int16" }
+            @{ Script = "0x10l"; Expected = "Int64" }
+            # No decimal hex literals (d is a valid hex char, and decimal parsing does not accept hex values)
+            @{ Script = "0x10d"; Expected = "Int32" }
+        )
+        It "<Script> should result in an [<Expected>] value" -TestCases $testLiteralTypes {
+            param($Script, $Expected)
+            ExecuteCommand $Script | Should -BeOfType $Expected
```
Comment (iSazonov) [inline test/powershell/Language/Parser/Parser.Tests.ps1]: I think it should be added to tests where we check values.

File: src/System.Management.Automation/engine/parser/tokenizer.cs [hunk 11]
```diff
@@ -3294,15 +3329,46 @@ private static bool TryGetNumberValue(string strNum, bool hex, bool real, char s
...
+                                    result = (ushort)u;
+                                }
+                                else if (suffix.HasFlag(NumberSuffixFlags.Long) || u > uint.MaxValue)
+                                {
+                                    result = u;
+                                }
+                                else
+                                {
+                                    result = (uint)u;
+                                }
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/parser/tokenizer.cs]: Maybe more clear:
```c#
                               else if (suffix.HasFlag(NumberSuffixFlags.Long) || u <= uint.MaxValue)  
                                {  
                                     result = (uint)u;
                                 }  
                                else  
                                {  
                                    // ulong
                                    result = u;
                                 }  
```

File: src/System.Management.Automation/engine/parser/tokenizer.cs (hunk 9 above)
Comment (vexx32) [reply src/System.Management.Automation/engine/parser/tokenizer.cs]: In what sense do you mean? Just an example?

File: test/powershell/Language/Parser/Parser.Tests.ps1 (hunk 10 above)
Comment (vexx32) [reply test/powershell/Language/Parser/Parser.Tests.ps1]: Hmm... Oh, you mean with those others there like the long checks...

I know what you mean, but after running into the quick with Int16 values I felt it perhaps might be a little more prudent to directly check types rather than checking the way some of the other tests 'check type'. It's hard to guarantee the value won't just be cast into the same form when it may not be quite correct.

File: src/System.Management.Automation/engine/parser/tokenizer.cs (hunk 9 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/parser/tokenizer.cs]: Yes, like `0x10d`.

File: test/powershell/Language/Parser/Parser.Tests.ps1 (hunk 10 above)
Comment (iSazonov) [reply test/powershell/Language/Parser/Parser.Tests.ps1]: We can test type before any operations on value.

Comment (vexx32) [reply test/powershell/Language/Parser/Parser.Tests.ps1]: I'm not following, I think. The existing test for values doesn't consider type except in a few cases, and even in those the type checks are implicit, not explicit, and casting may occur.

File: src/System.Management.Automation/engine/parser/tokenizer.cs (hunk 9 above)
Comment (vexx32) [reply src/System.Management.Automation/engine/parser/tokenizer.cs]: Will do! 😄

File: test/powershell/Language/Parser/Parser.Tests.ps1 (hunk 10 above)
Comment (iSazonov) [reply test/powershell/Language/Parser/Parser.Tests.ps1]: Can we do:
```powershell
	Context "Numerical Notations Tests (starting at line 2374 to line 2452)" {
        $testData = @(
			#Test various numbers using the standard notation.
            @{ Script = "0"; Expected = "0" }
			@{ Script = "-2"; Expected = "-2"; ExpectedType = [int32] }
			@{ Script = "2"; Expected = "2"; ExpectedType = [int32] }
			@{ Script = $([int32]::MaxValue); Expected = $([int32]::MaxValue); ExpectedType = [int32] }
			@{ Script = $([int32]::MinValue); Expected = $([int32]::MinValue); ExpectedType = [int32] }

...

        It "<Script> should return <Expected>" -TestCases $testData {
            param ( $Script, $Expected, $ExpectedType )
            $res = ExecuteCommand $Script
            $res | Should -Be $Expected
             $res | Should -BeType $ExpectedType
       }

File: src/System.Management.Automation/engine/parser/tokenizer.cs (hunk 11 above)
Comment (vexx32) [reply src/System.Management.Automation/engine/parser/tokenizer.cs]: Oh, you mean have the ulong case come last and alter the penultimate else if... Hmm. Yeah, OK, I can see the merit in that. Will do.

File: test/powershell/Language/Parser/Parser.Tests.ps1 (hunk 10 above)
Comment (vexx32) [reply test/powershell/Language/Parser/Parser.Tests.ps1]: Roll it all into one.... Hm. Yeap, should be able to. I'll check into it shortly.

File: src/System.Management.Automation/engine/parser/tokenizer.cs (hunk 11 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/parser/tokenizer.cs]: Closed.

File: src/System.Management.Automation/engine/parser/tokenizer.cs (hunk 9 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/parser/tokenizer.cs]: Closed.

File: test/powershell/Language/Parser/Parser.Tests.ps1 (hunk 10 above)
Comment (iSazonov) [reply test/powershell/Language/Parser/Parser.Tests.ps1]: If we continue in follow PR to add optinizations in the code and fix a hex sign it is important check not only parsing each string here but also test a result type. I think this will help a lot in follow PR avoid regressions. 
If you discover that we check here not all numeric formats please add new lines in the test.

File: test/powershell/Language/Parser/Parser.Tests.ps1 [hunk 12]
```diff
@@ -649,48 +649,72 @@ foo``u{2195}abc
...
+			@{ Script = "9.12"; Expected = "9.12"; ExpectedType = [double] }
+			@{ Script = $([single]::MinValue); Expected = $([float]::MinValue).ToString(); ExpectedType = [single] }
+			@{ Script = $([float]::MaxValue); Expected = $([float]::MaxValue).ToString(); ExpectedType = [float] }
 			#Tests for the K suffix for numbers.
-			@{ Script = "0kb"; Expected = "0" }
-			@{ Script = "1kb"; Expected = "1024" }
-			@{ Script = "-2KB"; Expected = "-2048" }
+			@{ Script = "0kb"; Expected = "0"; ExpectedType = [int] }
+			@{ Script = "1kb"; Expected = "1024"; ExpectedType = [int] }
+			@{ Script = "-2KB"; Expected = "-2048"; ExpectedType = [int] }
```
Comment (iSazonov) [inline test/powershell/Language/Parser/Parser.Tests.ps1]: We don't test all code paths with suffixes. Ex.: 9.12Kb, -2lkb. Could you please enhance tests?

I suggest remove lines starting with "#Tests for the K suffix for numbers.". Then add test code like we did for expected type tests.
```powershell
Context "Numerical Notations Tests (starting at line 2374 to line 2452)" {
    BeforeAll {
$suffixes = @(
     @{ Name = ""; Value = 1 }
    @{ Name = "kb"; Value = 1024 }
    @{ Name = "MB"; Value = 1024*1024 }
    @{ Name = "Gb"; Value = 1024*1024*1024 }
    @{ Name = "Pb"; Value = 1024*1024*1024*1024 }
)
}

It "<Script> should return <Expected>, with type <ExpectedType>" -TestCases $testData {  
     param ( $Script, $Expected, $ExpectedType )  

foreach ($suffix in $Suffixes) {
    $NumWithSuffix = $Script + $suffix.Name
    $ExpectedNum = $Expected * $suffix.Value
    ExecuteCommand $NumWithSuffix | Should -BeOfType $ExpectedType
    ExecuteCommand $NumWithSuffix | Should -Be $ExpectedNum
}
}

File: src/System.Management.Automation/engine/parser/tokenizer.cs [hunk 13]
```diff
@@ -3284,37 +3284,41 @@ private static bool TryGetNumberValue(string strNum, bool hex, bool real, Number
...
+                                    break;
                                 case NumberSuffixFlags.Long:
                                     result = ((long)Convert.ChangeType(d, typeof(long), CultureInfo.InvariantCulture) * multiplier);
                                     break;
                                 case NumberSuffixFlags.Short:
                                     result = (short)((short)Convert.ChangeType(d, typeof(short), CultureInfo.InvariantCulture) * multiplier);
                                     break;
-                                default:
-                                    if (suffix.HasFlag(NumberSuffixFlags.Unsigned))
+                                case NumberSuffixFlags.Unsigned | NumberSuffixFlags.Long:
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/parser/tokenizer.cs]: | -> && ?

File: src/System.Management.Automation/engine/parser/tokenizer.cs [hunk 14]
```diff
@@ -3294,15 +3334,47 @@ private static bool TryGetNumberValue(string strNum, bool hex, bool real, char s
...
+                    switch (suffix)
                     {
-                        if (long.TryParse(strNum, style, NumberFormatInfo.InvariantInfo, out longValue))
-                        {
-                            result = longValue * multiplier;
-                            return true;
-                        }
-                        result = null;
-                        return false;
+                        case NumberSuffixFlags.Long:
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/parser/tokenizer.cs]: What about "ul"?

Comment (vexx32) [reply src/System.Management.Automation/engine/parser/tokenizer.cs]: Presently, that one lives under the default case, inside the checks for the Unsigned flags.

Comment (iSazonov) [reply src/System.Management.Automation/engine/parser/tokenizer.cs]: Sorry case is not HasFlag. Closed.

File: test/powershell/Language/Parser/Parser.Tests.ps1 (hunk 12 above)
Comment (vexx32) [reply test/powershell/Language/Parser/Parser.Tests.ps1]: It's a good idea to do that, but since high values and high multipliers can result in overflow errors I think we're unfortunately a bit too likely to get failing tests by just iterating through the multipliers with everything (Pb overflows Int32, Gb overflows Int16, etc., etc.)

I've added a bunch, let me know if you think this is sufficient. 😄

**Comment by iSazonov** [comment]
@vexx32 Thanks for adding new test but they do not cover all code paths. Also we will add other code paths in the subsequent PR. That's why I suggested using a loop for suffixes.

**Comment by vexx32** [comment]
Hmm, I see your point there, but as mentioned that will cause its own issues. I'll comb through the code paths and put together some more thorough tests from the beginning, should be a bit simpler than trying to squint at all the tests and see what's missing. 😄

**Comment by iSazonov** [comment]
@vexx32 We pass a string in `ExecuteCommand` and call AddScript there - I don't see side effects. Then we directly return a result object - again we shouldn't catch side effects.

**Comment by vexx32** [comment]
Sorry, poor choice of words @iSazonov.

I'm thinking on it... I need to expand the tests and organise them a bit regardless, but as long as I don't add cases for the `short` or `ushort` type-d literal tests we should be fine. (Multipliers above `kb` will fail if applied to `(u)short` values because they are outside the value range of those data types.)

**Comment by iSazonov** [comment]
My understanding is that in ExecuteCommand a test string will be parsed. If we would wanted to test only parser we would not call ExecuteCommand at all. So we test not only parsing but conversion to value too.
As for multipliers above kb - if we expected a fail we should check it too. I think it is very useful if we plan add new custom ToInt()-like methods (without throw). It allow us to check all edge cases (overflows). We could add new parameter "ExpectedThrow" in the test.

**Comment by vexx32** [comment]
Hmm. Might be a bit janky determining programmatically which cases are meant to fail. I'll keep looking at it, but I think you might be on a good track there.

**Comment by vexx32** [comment]
@iSazonov So I did give what you recommended a shot, but I ran into a *lot* of issues trying to handle edge cases, where it should convert up to (u)long values, where it shouldn't convert up, where it should/shouldn't error.

I hope what I've got here is a happy medium between 'holy mother of test cases, Batman!' and 'what the hell is that weird loop doing'. 😉 

Let me know if you'd like me to expand it any; with all the edge cases on the numerals, handling it programmatically is liable to be just another painful piece of code to try to upkeep. At least with explicit declaration we know exactly what it's testing, why, and how. 😄

**Review by iSazonov — COMMENTED** [review]
@vexx32 We also support '+' on first place in numerics. I think we should add such tests too.

**Comment by vexx32** [comment]
Added! 😄 

And fixed those real `0.0` cases that I apparently didn't notice.

**Review by iSazonov — APPROVED** [review]

**Comment by iSazonov** [comment]
@rjmholt @SteveL-MSFT @daxian-dbw @BrucePay Please review the PR.

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee re-reviewed this and maintains the same recommendation to move forward with this PR

**Comment by vexx32** [comment]
Hmm, haven't seen *that* error yet. Looks like an unrelated timeout issue, though. 😄

**Comment by iSazonov** [comment]
@vexx32 Please rebase to fix CIs.

**Comment by vexx32** [comment]
Attempting... this always seems far messier than it should be.

**Comment by vexx32** [comment]
Okay... @iSazonov apologies about the mess. Not sure what it is with me and rebasing, but I've yet to manage to figure it out and get it done cleanly. The Linux build seems to be failing on some kind of permissions issue with that build environment, not sure what that's about.

Everything else looks green, though. :)

**Comment by iSazonov** [comment]
You should reset to your last commit and then rebase.
Or create new branch, copy your commits and open new PR.

**Comment by vexx32** [comment]
@iSazonov yeah, point taken. Rebasing is something I've yet to get to grips with. CLosing this for now, let's continue from a cleaner slate in #7813, where everything's all ready. :D
